hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/TableRecordReaderImpl.java

```diff
@@ -66,12 +76,18 @@ public class TableRecordReaderImpl {
   private boolean logScannerActivity = false;
   private int logPerRowCount = 100;
 
+  private Duration renewLeasePeriod = null;
+
+  private Instant lastScannerRpcAt = null;
+
+  private ScheduledExecutorService renewTaskExecutor = null;
+
   /**
    * Restart from survivable exceptions by creating a new scanner.
    * @param firstRow The first row to start at.
    * @throws IOException When restarting fails.
    */
-  public void restart(byte[] firstRow) throws IOException {
+  public synchronized void restart(byte[] firstRow) throws IOException {
```

Comment:
these 3 methods are synchronized because it probably would not be good if renewLease were allowed to execute while creating the scanner or calling `scanner.next`.